### PR TITLE
birthday

### PR DESCRIPTION
**Pull request by nathanleiby**
Issue: https://clever.atlassian.net/browse/INFRA-1405

See README for an overview of `log-monitor`.

PR with passing test cases here, which include go vendoring: https://github.com/Clever/log-monitor/pull/2

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	"strconv"
+	"time"
+
+	"github.com/signalfx/golib/datapoint"
+	"github.com/signalfx/golib/sfxclient"
+	"golang.org/x/net/context"
+)
+
+func main() {
+	path := "log.txt"
```
Comment (burnsed) [inline main.go]: Should this be an argument or configurable via an env var?

File: main.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// getTimestampFromLog takes an Rsyslog logline and returns a timestamp
+// TODO: Consider using second-level timestamps. < second is too fine grained for this use-case.
+func getTimestampFromLog(s string) (int64, error) {
+	// Pull date from log line
+	const rsyslogFormat = "2006 Jan 2 15:04:05"
+	// Strip off the first 15 chars, which aligns with the `format` above (except missing the year)
+	if len(s) < 15 {
+		return 0, fmt.Errorf("log line < 15 chars, cannot contain date substring")
```
Comment (burnsed) [inline main.go]: does this mean that the log line does not contain a date?

Comment (nathanleiby) [inline main.go]: It may have a date embedded, but it's not in the first 15 chars (which is our Rsyslog syntax)

File: main.go (hunk 1 above)
Comment (nathanleiby) [inline main.go]: yup! didn't quite add that yet

File: README.md [hunk 3]
```diff
@@ -1,2 +1,21 @@
...
-Verifies logs are forwarded successfully / in a timely manner.
+
+Run on several key logging machines which have log forwarding configured.
+Periodically sends the timestamp of the latest log to SignalFX.
+We can then compare timestamps from different hosts and verify that logs were forwaded successfully and quickly.
+
+## Usage
+
+Configure an environment variable `SIGNALFX_API_KEY` with your SignalFX API Key.
+
```
Comment (burnsed) [inline README.md]: might be good to add a section describing the format it expects for the log lines

File: main.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		return 0, nil
+	}
+
+	// Infer the same year as `now`, unless we've crossed the EOY boundary.
+	year := now.Year()
+	if now.Month() == time.January && t.Month() == time.December {
+		year--
+	}
+
+	return year, nil
```
Comment (burnsed) [inline main.go]: :+1:

**Comment by burnsed** [comment]
some small nits, but otherwise LGTM!

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,46 @@
+.PHONY: build run vendor
```
Comment (mohit) [inline Makefile]: use the go-guild go Makefile? https://github.com/Clever/dev-handbook/tree/master/make#golang

File: main.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	return out.String(), nil
+}
+
+// getTimestampFromLog takes an Rsyslog logline and returns a timestamp
+// TODO: Consider using second-level timestamps. < second is too fine grained for this use-case.
+func getTimestampFromLog(s string) (int64, error) {
+	// Pull date from log line
+	const rsyslogFormat = "2006 Jan 2 15:04:05"
```
Comment (mohit) [inline main.go]: also maybe a config parameter?

File: main.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+	err = client.AddDatapoints(context.TODO(), points)
+	if err != nil {
+		log.Println("error writing datapoint to SignalFx:", err)
+		return
+	}
+}
+
+func getMostRecentLog(p string) (string, error) {
+	cmd := exec.Command("tail", "-1", p)
```
Comment (mohit) [inline main.go]: just a nit about using a syscall. Could use `Seek`. ala https://github.com/coreymgilmore/gotail/blob/master/gotail.go#L56

File: main.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+	t, err := getTimestampFromLog(s)
+	if err != nil {
+		log.Println("error getting timestamp from log:", err)
+		log.Println("log:", s)
+		return
+	}
+
+	client := sfxclient.NewHTTPDatapointSink()
+	client.AuthToken = os.Getenv("SIGNALFX_AUTH_TOKEN")
```
Comment (mohit) [inline main.go]: fail early (in `main`) if this is not present?

Comment (nathanleiby) [inline main.go]: Fail early: 2621c16542fc79cbed5b0645757813852d9edc9c

File: README.md (hunk 3 above)
Comment (nathanleiby) [inline README.md]: Add log syntax to readme: https://github.com/Clever/log-monitor/commit/a89c19476893f645f61ea35aa9135679fb7929fc

File: main.go (hunk 6 above)
Comment (nathanleiby) [inline main.go]: The `inferYear` logic makes this a little harder to do. I can add it as a feature if useful later.

File: Makefile (hunk 5 above)
Comment (nathanleiby) [inline Makefile]: Added go-guild makefile: e641a59

File: main.go (hunk 1 above)
Comment (nathanleiby) [inline main.go]: Allow passing `path` to logfile eacf900

File: Makefile (hunk 5 above)
Comment (natebrennand) [inline Makefile]: :clap:

File: main.go (hunk 7 above)
Comment (nathanleiby) [inline main.go]: Thanks! I did consider this approach, but thought syscall was simplest at the moment. 

Do you prefer golang native solution for portability, safety, and/or another reason?

**Pull request merged**